library/alloc/src/boxed/thin.rs

```diff
@@ -91,6 +93,83 @@ impl<T> ThinBox<T> {
 
 #[unstable(feature = "thin_box", issue = "92791")]
 impl<Dyn: ?Sized> ThinBox<Dyn> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new_unsize_zst<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        #[repr(C)]
+        struct ReprC<A, B> {
+            a: A,
+            b: B,
+        }
+
+        // Allocate header like this:
+        // ```
+        // [ ... | header ]
+        // ```
+        // where the struct is aligned to both header and value.
+        #[repr(C)]
+        struct AlignedHeader<H: Copy, T> {
+            header_data: MaybeUninit<ReprC<H, [T; 0]>>,
+        }
+
+        impl<H: Copy, T> AlignedHeader<H, T> {
+            const fn make(header: H) -> Self {
+                let mut data = MaybeUninit::<ReprC<H, [T; 0]>>::zeroed();
+                unsafe {
+                    data.as_mut_ptr().add(1).cast::<H>().sub(1).write(header);
+                }
+                AlignedHeader { header_data: data }
+            }
+        }
+
+        #[repr(C)]
+        struct DynZstAlloc<T, Dyn: ?Sized> {
+            header: AlignedHeader<<Dyn as Pointee>::Metadata, T>,
+            value: MaybeUninit<T>,
+        }
+
+        // We need `Freeze` so we could call `&DynZstAlloc::LAYOUT`.
+        // SAFETY: data is immutable.
+        unsafe impl<T, Dyn: ?Sized> Freeze for DynZstAlloc<T, Dyn> {}
```

Comment:
There should probably be a `Freeze` bound on the definition of [`Pointee::Metadata`](https://doc.rust-lang.org/nightly/core/ptr/trait.Pointee.html#associatedtype.Metadata), otherwise the blanket impls of `impl<T: ?Sized> Freeze for &T` would be unsound if we ever add custom DSTs with `!Freeze` metadata.

Reply:
https://github.com/rust-lang/rust/pull/123206

Reply:
Is it possible to use `repr(align = ...)` instead?

Reply:
`repr(align(...))` requires integer literal argument. Here we need to align to `T`.